lib/factorio/server.js

```diff
@@ -638,7 +647,33 @@ class FactorioServer extends events.EventEmitter {
 		// developers.  Rcon does not recognize /quit, stdin is not
 		// recognized, and there's no "send CTRL+C to process" on Windows.
 		this._server.kill();
-		await events.once(this._server, 'exit');
+
+		// There appears to be an race condition where sending SIGTERM
+		// immediatly before the RCON interface comes online causes the
+		// Factorio server to hang.  It's also possible the server doesn't
+		// stop if it's stuck with an infinite lua code loop.
+		if (process.platform !== "win32") {
+			let hanged = true;
+			function setAlive(output) { hanged = false; }
+			this.on('_quitting', setAlive);
+
+			let timeoutId = setTimeout(() => {
+				if (hanged) {
+					console.error("Factorio appears to have hanged, sending second SIGTERM");
+					this._server.kill();
+				}
+			}, 5000);
```

Comment:
I don't think 5 seconds is really enough. Bigger maps can easily take much more to save, we have seen up to 20-25 seconds.

Reply:
This timeout is from sending SIGTERM until receiving the "Quitting: signal" message on stdout, which normally take no more that a millisecond.  Map saving happens after that so this code will wait indefinitely for saving to complete.